solidity/contracts/bridge/WalletCoordinator.sol

```diff
@@ -0,0 +1,571 @@
+    uint32 public depositSweepProposalValidity;
+
+    /// @notice The minimum time that must elapse since the deposit reveal
+    ///         before a deposit becomes eligible for a deposit sweep.
+    ///
+    ///         For example, if a deposit was revealed at 9 am and depositMinAge
+    ///         is 2 hours, the deposit is eligible for sweep after 11 am.
+    uint32 public depositMinAge;
+
+    /// @notice Each deposit can be technically swept until it reaches its
+    ///         refund timestamp after which it can be taken back by the depositor.
+    ///         However, allowing the wallet to sweep deposits that are close
+    ///         to their refund timestamp may cause a race between the wallet
+    ///         and the depositor. In result, the wallet may sign an invalid
+    ///         sweep transaction that aims to sweep an already refunded deposit.
+    ///         Such tx signature may be used to create an undefeatable fraud
+    ///         challenge against the wallet. In order to mitigate that problem,
+    ///         this parameter determines a safety margin that puts the latest
+    ///         moment a deposit can be swept far before the point after which
+    ///         the deposit becomes refundable.
+    ///
+    ///         For example, if a deposit becomes refundable after 8 pm and
+    ///         depositRefundSafetyMargin is 6 hours, the deposit is valid for
+    ///         for a sweep only before 2 pm.
+    uint32 public depositRefundSafetyMargin;
+
+    /// @notice The maximum count of deposits that can be swept within a
+    ///         single sweep.
+    uint16 public depositSweepMaxSize;
+
+    event ProposalSubmitterAdded(address indexed proposalSubmitter);
+
+    event ProposalSubmitterRemoved(address indexed proposalSubmitter);
+
+    event WalletManuallyUnlocked(bytes20 indexed walletPubKeyHash);
+
+    event DepositSweepProposalValidityUpdated(
+        uint32 depositSweepProposalValidity
+    );
+
+    event DepositMinAgeUpdated(uint32 depositMinAge);
+
+    event DepositRefundSafetyMarginUpdated(uint32 depositRefundSafetyMargin);
+
+    event DepositSweepMaxSizeUpdated(uint16 depositSweepMaxSize);
+
+    event DepositSweepProposalSubmitted(
+        DepositSweepProposal proposal,
+        address indexed proposalSubmitter
+    );
+
+    modifier onlyProposalSubmitterOrWalletMember(
+        bytes20 walletPubKeyHash,
+        WalletMemberContext calldata walletMemberContext
+    ) {
+        bool proposalSubmitter = isProposalSubmitter[msg.sender];
+        bool walletMember = false;
+
+        if (!proposalSubmitter) {
+            bytes32 ecdsaWalletID = bridge
+                .wallets(walletPubKeyHash)
+                .ecdsaWalletID;
+
+            walletMember = walletRegistry.isWalletMember(
+                ecdsaWalletID,
+                walletMemberContext.walletMembersIDs,
+                msg.sender,
+                walletMemberContext.walletMemberIndex
+            );
+        }
```

Comment:
How gas-expensive is the `if (!proposalSubmitter)` path compared to the `proposalSubmitter == true` path?

Reply:
The `proposalSubmitter == true` is `44401` on average. The `if !(proposalSubmitter)` path is `65338` on average.

Reply:
OK, so it's at least 30% of the cost of calling this function if called by the coordinator. Let's keep in mind when addressing this [TODO](https://github.com/keep-network/tbtc-v2/pull/575#discussion_r1151564813) that it has to be gas-optimized solution.